sys/include/net/sock/ip.h

```diff
@@ -527,6 +530,49 @@ static inline ssize_t sock_ip_recv(sock_ip_t *sock, void *data, size_t max_len,
     return sock_ip_recv_aux(sock, data, max_len, timeout, remote, NULL);
 }
 
+#if IS_USED(MODULE_SOCK_PEEK) || defined(DOXYGEN)
+/**
+ * @brief   Peeks bytes from a message over IPv4/IPv6 from remote end point. If
+ *          no data is available then an attempt is made to receive new data first.
+ *
+ * @note    Currently only available for lwIP!
+ *
+ * @pre `(sock != NULL) && (data != NULL) && (max_len > 0)`
+ *
+ * @param[in] sock      A raw IPv4/IPv6 sock object.
+ * @param[out] data     Pointer where the read data should be stored.
+ * @param[in] max_len   Maximum space available at @p data.
+ * @param[in] timeout   Timeout for receive in microseconds.
+ *                      If 0 and no data is available, the function returns
+ *                      immediately.
+ *                      May be @ref SOCK_NO_TIMEOUT for no timeout (wait until
+ *                      data is available).
+ * @param[out] remote   Remote end point of the received data.
+ *                      May be NULL, if it is not required by the application.
+ *
+ * @note    Function blocks if no packet is currently waiting.
+ * @note    This function is only available when using lwIP at the moment.
+ *
+ * @return  The number of bytes read on success.
+ * @return  0, if no read data is available, but everything is in order.
+ * @return  -EADDRNOTAVAIL, if local of @p sock is not given.
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
+ * @return  -EINVAL, if @p remote is invalid or @p sock is not properly
+ *          initialized (or closed while sock_ip_recv() blocks).
+ * @return  -ENOMEM, if no memory was available to receive @p data.
+ * @return  -EPROTO, if source address of received packet did not equal
+ *          the remote of @p sock.
+ * @return  -ETIMEDOUT, if @p timeout expired.
+ */
+static inline ssize_t sock_ip_peek(sock_ip_t *sock, void *data, size_t max_len,
```

Comment:
For TCP, an explicit function sounds viable -- having an aux there would probably be good at some point (MSG_MORE comes to mind), but I have no strong ideas about whether that's better done now or later. (For when it is done later, a comment a la `/* Compared to other peeking sockets, this is done in a dedicated function and not sock_tcp_read_aux because there is no such function; if one is ever added, consider making "peek" a flag there and deprecating this function */` would capture the roadmap on my mind).

Reply:
Only TCP has a dedicated function now